### PR TITLE
fixes #3018 - API v2 metadata for pagination, search, order in index collection json response

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by isratrade** [comment]
@komidore64, @domcleal, here are some examples.  

Collections example:
I searched for the letter 't' in my locations and want the results in descending order.

0.0.0.0:3000/api/locations?search=t&order=name+DESC

```
{
    "total": 28,
    "subtotal": 3,
    "limit": 3,
    "offset": 0,
    "page": 1,
    "per_page": 20,
    "search": "t",
    "sort": {
        "by": "name",
        "order": "DESC"
    },
    "results": [
        {
            "id": 34,
            "name": "tt"
        },
        {
            "id": 3,
            "name": "TLV"
        },
        {
            "id": 10,
            "name": "The Cloud"
        }
    ]
}
```

**Comment by isratrade** [comment]
limit/offset and page/per_page might be a little redundant.  the background is that katello uses offset/limit and foreman uses page/per_page.

**Comment by isratrade** [comment]
Single object example.
pretty straightforward
- node name is object type, so it varies as opposed to collections which always return 'results'
- no meta data

0.0.0.0:3000/api/environments/production

```
{
    "environment": {
        "name": "production",
        "id": 1,
        "created_at": "2010-08-15T10:24:09Z",
        "updated_at": "2010-08-15T10:24:09Z"
    }
}
```

**Comment by domcleal** [comment]
get_metadata is only called from the taxonomies controller, isn't it needed in the V2 base controller too?

**Comment by isratrade** [comment]
I wanted to call `get_metadata(results)` in an `after_filter`, but there's a problem with this since I need to get the metadata instance variables after @taxonomies is defined but before `render 'api/v2/taxonomies/index'` is called.  Unless I fine a way to call it before render, I need to put `get_metadata(results)` in each individual controller.  Since this PR only uses taxonomies_controller as the first testcase, it only appears there.

**Comment by isratrade** [comment]
added gem rails3_before_render and refactored code in version2.rb to v2/base_controller.rb
closing PR #862 so I woudn't need to rebase two PR's

File: app/controllers/api/v2/puppetclasses_controller.rb [hunk 1]
```diff
@@ -29,7 +30,7 @@ def index
                      else
                        nested_obj.puppetclasses.search_for(*search_options)
                    end
-        render :json => Puppetclass.classes2hash(values.paginate(paginate_options))
+        render :json => {:results => Puppetclass.classes2hash(values.paginate(paginate_options))}
```
Comment (isratrade) [inline app/controllers/api/v2/puppetclasses_controller.rb]: root node for puppetclasses is now :results

**Comment by domcleal** [comment]
The only thing that concerns me about merging this is that we haven't implemented #3017 yet, so the APIv2 is going to be inconsistent up til that point.. I'd prefer to do it the other way around.

The other thing is I've updated the packaging for the new dependency, so this will need squashing in on merge: https://github.com/domcleal/foreman/commit/ffc93cdd37a67b12c0ece82e6c276ef9e4212367.  [Tested via Jenkins](http://ci.theforeman.org/job/packaging_build_matrix_rpm/373/).

**Comment by isratrade** [comment]
@domcleal, no problem.  I'll add a commit for #3017 to this PR.

**Comment by isratrade** [comment]
@domcleal, all api PR's are re-based and ready for merge after your review.  fyi, @thomasmckay

**Comment by komidore64** [comment]
will these changes be easily consumable by an engine?

**Comment by isratrade** [comment]
@komidore64, I don't understand the question.  The bastion / angular app (which is an engine) is consuming the Foreman API v2, but any api client can consume it.

**Comment by komidore64** [comment]
@isratrade my apologies, i worded badly. will these changes take effect in engines as well, (i.e. - will katello get this metadata simply by being an engine of foreman?) or will engines need their own metadata code?

----- Original Message -----

> From: "Joseph Magen" notifications@github.com
> To: "theforeman/foreman" foreman@noreply.github.com
> Cc: "Adam Price" adprice@redhat.com
> Sent: Monday, November 25, 2013 2:37:25 AM
> Subject: Re: [foreman] fixes #3018 - API v2 metadata for pagination, search, order in index collection json response
> (#952)
> 
> @komidore64, I don't understand the question.  The bastion / angular app
> (which is an engine) is consuming the Foreman API v2, but any api client can
> consume it.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/theforeman/foreman/pull/952#issuecomment-29182359

## 
- adam price

**Comment by isratrade** [comment]
@komidore64, engines can use views from the main_app but isolated engines usually use views from their own.  You just need to pass `:layout =>` in the render call.

**Comment by isratrade** [comment]
@domcleal, tasks_controller modified to add `results` node

File: app/controllers/api/v2/base_controller.rb [hunk 2]
```diff
@@ -21,6 +22,32 @@ def root_node_name
                           end
       end
 
+      def get_metadata
+        @results ||= instance_variable_get("@#{controller_name}")
+        #@total should be defined in individual controllers, but in case it's not.
+        @total ||= @results.try(:count).to_i
```
Comment (domcleal) [inline app/controllers/api/v2/base_controller.rb:28]: Will this be the total after pagination on most controllers?

Comment (domcleal) [inline app/controllers/api/v2/base_controller.rb:28]: Oh no, ignore me, it's the actual total irrespective of pagination.

File: app/controllers/api/v2/puppetclasses_controller.rb [hunk 3]
```diff
@@ -7,6 +7,7 @@ class PuppetclassesController < V2::BaseController
 
       before_filter :find_resource, :only => %w{show update destroy}
       before_filter :find_optional_nested_object, :only => [:index]
+      skip_before_render :get_metadata, :only => :index
```
Comment (domcleal) [inline app/controllers/api/v2/puppetclasses_controller.rb]: Can we provide metadata manually in index?

**Comment by domcleal** [comment]
Merged as 92b7b44, thanks @isratrade!

**Pull request closed without merging**